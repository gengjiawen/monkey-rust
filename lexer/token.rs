use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Formatter;

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "start: {}, end: {}, kind: {}", self.span.start, self.span.end, self.kind)
    }
}

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
#[serde(tag = "type", content = "value")]
pub enum TokenKind {
    ILLEGAL,
    EOF,

    // Identifiers + literals
    IDENTIFIER { name: String },
    INT(i64),
    STRING(String),

    // Operators
    ASSIGN,   // =
    PLUS,     // +
    MINUS,    // -
    BANG,     // !
    ASTERISK, // *
    SLASH,    // /

    LT, // <
    GT, // >

    EQ,    // ==
    NotEq, // !=

    // delimiters
    COMMA,
    SEMICOLON,
    COLON,

    LPAREN,
    RPAREN,
    LBRACE,   // {
    RBRACE,   // }
    LBRACKET, // [
    RBRACKET, // ]

    // keywords
    FUNCTION,
    LET,
    TRUE,
    FALSE,
    IF,
    ELSE,
    RETURN,
}

pub fn lookup_identifier(identifier: &str) -> TokenKind {
    match identifier {
        "fn" => TokenKind::FUNCTION,
        "let" => TokenKind::LET,
        "true" => TokenKind::TRUE,
        "false" => TokenKind::FALSE,
        "if" => TokenKind::IF,
        "else" => TokenKind::ELSE,
        "return" => TokenKind::RETURN,
        _ => TokenKind::IDENTIFIER { name: identifier.to_string() },
    }
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            TokenKind::IDENTIFIER { name } => write!(f, "{}", name),
            TokenKind::INT(i) => write!(f, "{}", i),
            TokenKind::STRING(s) => write!(f, "{}", s),
            TokenKind::ASSIGN => write!(f, "="),
            TokenKind::PLUS => write!(f, "+"),
            TokenKind::MINUS => write!(f, "-"),
            TokenKind::BANG => write!(f, "!"),
            TokenKind::ASTERISK => write!(f, "*"),
            TokenKind::SLASH => write!(f, "/"),
            TokenKind::LT => write!(f, "<"),
            TokenKind::GT => write!(f, ">"),
            TokenKind::EQ => write!(f, "=="),
            TokenKind::NotEq => write!(f, "!="),
            TokenKind::COMMA => write!(f, ","),
            TokenKind::SEMICOLON => write!(f, ";"),
            TokenKind::LPAREN => write!(f, "("),
            TokenKind::RPAREN => write!(f, ")"),
            TokenKind::LBRACE => write!(f, "{{"),
            TokenKind::RBRACE => write!(f, "}}"),
            TokenKind::LBRACKET => write!(f, "["),
            TokenKind::RBRACKET => write!(f, "]"),
            TokenKind::FUNCTION => write!(f, "fn"),
            TokenKind::LET => write!(f, "let"),
            TokenKind::TRUE => write!(f, "true"),
            TokenKind::FALSE => write!(f, "false"),
            TokenKind::IF => write!(f, "if"),
            TokenKind::ELSE => write!(f, "else"),
            TokenKind::RETURN => write!(f, "return"),
            TokenKind::ILLEGAL => write!(f, "ILLEGAL"),
            TokenKind::EOF => write!(f, "EOF"),
            TokenKind::COLON => write!(f, ":"),
        }
    }
}
