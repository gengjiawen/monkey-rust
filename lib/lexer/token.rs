
#[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
pub enum Token {
    ILLEGAL,
    EOF,

    // Identifiers + literals
    IDENT(String),
    Int(i64),

    // Operators
    ASSIGN,
    PLUS,

    // delimiters
    COMMA,
    SEMICOLON,

    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,

    // keywords
    FUNCTION,
    LET
}

