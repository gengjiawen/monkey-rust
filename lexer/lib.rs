use crate::token::{lookup_identifier, Span, Token, TokenKind};

mod lexer_test;
pub mod token;

pub struct Lexer<'a> {
    input: &'a str,
    position: usize,
    read_position: usize,
    ch: char,
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut l = Lexer {
            input,
            position: 0,
            read_position: 0,
            ch: 0 as char,
        };

        l.read_char();
        return l;
    }

    fn read_char(&mut self) {
        if self.read_position >= self.input.len() {
            self.ch = 0 as char
        } else {
            if let Some(ch) = self.input.chars().nth(self.read_position) {
                self.ch = ch;
            } else {
                panic!("read out of range")
            }
        }

        self.position = self.read_position;
        self.read_position += 1;
    }

    fn peek_char(&self) -> char {
        if self.read_position >= self.input.len() {
            0 as char
        } else {
            if let Some(ch) = self.input.chars().nth(self.read_position) {
                ch
            } else {
                panic!("read out of range")
            }
        }
    }

    pub fn next_token(&mut self) -> Token {
        // println!("self ch {}, position {} read_position {}", self.ch, self.position, self.read_position);
        self.skip_whitespace();
        self.skip_comments();
        let t = match self.ch {
            '=' => {
                if self.peek_char() == '=' {
                    self.read_char();
                    TokenKind::EQ
                } else {
                    TokenKind::ASSIGN
                }
            }
            ';' => TokenKind::SEMICOLON,
            '(' => TokenKind::LPAREN,
            ')' => TokenKind::RPAREN,
            ',' => TokenKind::COMMA,
            '+' => TokenKind::PLUS,
            '-' => TokenKind::MINUS,
            '!' => {
                if self.peek_char() == '=' {
                    self.read_char();
                    TokenKind::NotEq
                } else {
                    TokenKind::BANG
                }
            }
            '*' => TokenKind::ASTERISK,
            '/' => TokenKind::SLASH,
            '<' => TokenKind::LT,
            '>' => TokenKind::GT,
            '{' => TokenKind::LBRACE,
            '}' => TokenKind::RBRACE,
            '[' => TokenKind::LBRACKET,
            ':' => TokenKind::COLON,
            ']' => TokenKind::RBRACKET,
            '\u{0}' => TokenKind::EOF,
            '"' => {
                let (start, end, string) = self.read_string();
                return Token {
                    span: Span { start, end },
                    kind: TokenKind::STRING(string),
                };
            }
            _ => {
                if is_letter(self.ch) {
                    let (start, end, identifier) = self.read_identifier();
                    return Token {
                        span: Span { start, end },
                        kind: lookup_identifier(&identifier),
                    };
                } else if is_digit(self.ch) {
                    let (start, end, num) = self.read_number();
                    return Token {
                        span: Span { start, end },
                        kind: TokenKind::INT(num),
                    };
                } else {
                    TokenKind::ILLEGAL
                }
            }
        };

        self.read_char();
        return Token {
            span: Span {
                start: self.position - 1,
                end: self.read_position - 1,
            },
            kind: t,
        };
    }

    fn skip_whitespace(&mut self) {
        while self.ch.is_ascii_whitespace() {
            self.read_char();
        }
    }

    fn skip_comments(&mut self) {
        if self.ch == '/' && self.peek_char() == '/' {
            self.read_char();
            self.read_char();
            loop {
                self.read_char();
                if self.ch == '\n' || self.ch == '\u{0}' {
                    // consume the comments end
                    if self.ch == '\n' {
                        self.read_char();
                    }
                    break;
                }
            }
        }
    }

    fn read_identifier(&mut self) -> (usize, usize, String) {
        let pos = self.position;
        while is_letter(self.ch) {
            self.read_char();
        }

        let x = self.input[pos..self.position].to_string();
        return (pos, self.position, x);
    }

    fn read_number(&mut self) -> (usize, usize, i64) {
        let pos = self.position;
        while is_digit(self.ch) {
            self.read_char();
        }

        let x = self.input[pos..self.position].parse().unwrap();

        return (pos, self.position, x);
    }

    fn read_string(&mut self) -> (usize, usize, String) {
        let pos = self.position + 1;
        loop {
            self.read_char();
            if self.ch == '"' || self.ch == '\u{0}' {
                break;
            }
        }

        let x = self.input[pos..self.position].to_string();

        // consume the end "
        if self.ch == '"' {
            self.read_char();
        }
        return (pos - 1, self.position, x);
    }
}

fn is_letter(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_digit(c: char) -> bool {
    c >= '0' && c <= '9'
}
