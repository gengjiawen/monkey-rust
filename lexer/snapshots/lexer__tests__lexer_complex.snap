---
source: lexer/lib.rs
expression: token_vs
---
[
    Token {
        start: 0,
        end: 3,
        kind: LET,
    },
    Token {
        start: 4,
        end: 8,
        kind: IDENTIFIER {
            name: "five",
        },
    },
    Token {
        start: 9,
        end: 10,
        kind: ASSIGN,
    },
    Token {
        start: 11,
        end: 12,
        kind: INT(
            5,
        ),
    },
    Token {
        start: 12,
        end: 13,
        kind: SEMICOLON,
    },
    Token {
        start: 14,
        end: 17,
        kind: LET,
    },
    Token {
        start: 18,
        end: 21,
        kind: IDENTIFIER {
            name: "ten",
        },
    },
    Token {
        start: 22,
        end: 23,
        kind: ASSIGN,
    },
    Token {
        start: 24,
        end: 26,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 26,
        end: 27,
        kind: SEMICOLON,
    },
    Token {
        start: 29,
        end: 32,
        kind: LET,
    },
    Token {
        start: 33,
        end: 36,
        kind: IDENTIFIER {
            name: "add",
        },
    },
    Token {
        start: 37,
        end: 38,
        kind: ASSIGN,
    },
    Token {
        start: 39,
        end: 41,
        kind: FUNCTION,
    },
    Token {
        start: 41,
        end: 42,
        kind: LPAREN,
    },
    Token {
        start: 42,
        end: 43,
        kind: IDENTIFIER {
            name: "x",
        },
    },
    Token {
        start: 43,
        end: 44,
        kind: COMMA,
    },
    Token {
        start: 45,
        end: 46,
        kind: IDENTIFIER {
            name: "y",
        },
    },
    Token {
        start: 46,
        end: 47,
        kind: RPAREN,
    },
    Token {
        start: 48,
        end: 49,
        kind: LBRACE,
    },
    Token {
        start: 52,
        end: 53,
        kind: IDENTIFIER {
            name: "x",
        },
    },
    Token {
        start: 54,
        end: 55,
        kind: PLUS,
    },
    Token {
        start: 56,
        end: 57,
        kind: IDENTIFIER {
            name: "y",
        },
    },
    Token {
        start: 57,
        end: 58,
        kind: SEMICOLON,
    },
    Token {
        start: 59,
        end: 60,
        kind: RBRACE,
    },
    Token {
        start: 60,
        end: 61,
        kind: SEMICOLON,
    },
    Token {
        start: 63,
        end: 66,
        kind: LET,
    },
    Token {
        start: 67,
        end: 73,
        kind: IDENTIFIER {
            name: "result",
        },
    },
    Token {
        start: 74,
        end: 75,
        kind: ASSIGN,
    },
    Token {
        start: 76,
        end: 79,
        kind: IDENTIFIER {
            name: "add",
        },
    },
    Token {
        start: 79,
        end: 80,
        kind: LPAREN,
    },
    Token {
        start: 80,
        end: 84,
        kind: IDENTIFIER {
            name: "five",
        },
    },
    Token {
        start: 84,
        end: 85,
        kind: COMMA,
    },
    Token {
        start: 86,
        end: 89,
        kind: IDENTIFIER {
            name: "ten",
        },
    },
    Token {
        start: 89,
        end: 90,
        kind: RPAREN,
    },
    Token {
        start: 90,
        end: 91,
        kind: SEMICOLON,
    },
    Token {
        start: 92,
        end: 93,
        kind: BANG,
    },
    Token {
        start: 93,
        end: 94,
        kind: MINUS,
    },
    Token {
        start: 94,
        end: 95,
        kind: SLASH,
    },
    Token {
        start: 95,
        end: 96,
        kind: ASTERISK,
    },
    Token {
        start: 96,
        end: 97,
        kind: INT(
            5,
        ),
    },
    Token {
        start: 97,
        end: 98,
        kind: SEMICOLON,
    },
    Token {
        start: 99,
        end: 100,
        kind: INT(
            5,
        ),
    },
    Token {
        start: 101,
        end: 102,
        kind: LT,
    },
    Token {
        start: 103,
        end: 105,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 106,
        end: 107,
        kind: GT,
    },
    Token {
        start: 108,
        end: 109,
        kind: INT(
            5,
        ),
    },
    Token {
        start: 109,
        end: 110,
        kind: SEMICOLON,
    },
    Token {
        start: 112,
        end: 114,
        kind: IF,
    },
    Token {
        start: 115,
        end: 116,
        kind: LPAREN,
    },
    Token {
        start: 116,
        end: 117,
        kind: INT(
            5,
        ),
    },
    Token {
        start: 118,
        end: 119,
        kind: LT,
    },
    Token {
        start: 120,
        end: 122,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 122,
        end: 123,
        kind: RPAREN,
    },
    Token {
        start: 124,
        end: 125,
        kind: LBRACE,
    },
    Token {
        start: 127,
        end: 133,
        kind: RETURN,
    },
    Token {
        start: 134,
        end: 138,
        kind: TRUE,
    },
    Token {
        start: 138,
        end: 139,
        kind: SEMICOLON,
    },
    Token {
        start: 140,
        end: 141,
        kind: RBRACE,
    },
    Token {
        start: 142,
        end: 146,
        kind: ELSE,
    },
    Token {
        start: 147,
        end: 148,
        kind: LBRACE,
    },
    Token {
        start: 150,
        end: 156,
        kind: RETURN,
    },
    Token {
        start: 157,
        end: 162,
        kind: FALSE,
    },
    Token {
        start: 162,
        end: 163,
        kind: SEMICOLON,
    },
    Token {
        start: 164,
        end: 165,
        kind: RBRACE,
    },
    Token {
        start: 167,
        end: 169,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 171,
        end: 172,
        kind: EQ,
    },
    Token {
        start: 173,
        end: 175,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 175,
        end: 176,
        kind: SEMICOLON,
    },
    Token {
        start: 177,
        end: 179,
        kind: INT(
            10,
        ),
    },
    Token {
        start: 181,
        end: 182,
        kind: NotEq,
    },
    Token {
        start: 183,
        end: 184,
        kind: INT(
            9,
        ),
    },
    Token {
        start: 184,
        end: 185,
        kind: SEMICOLON,
    },
    Token {
        start: 185,
        end: 186,
        kind: EOF,
    },
]
